### PR TITLE
Move to rspec3

### PR DESCRIPTION
**Pull request by electrical**
Ran the tests via `bundle exec rake` and all green.

### CODE REVIEW

**Comment by electrical** [comment]
@jeffmccune @rodjek any idea if this could get merged?

**Comment by logicminds** [comment]
@jeffmccune @rodjek any way this can be merged like now?  I would really like to use some of the new features in 2.0.0 but I use rspec 3 and this is blocking me.  Lets change it back to what it used to be before this mess started.

Without this setting we won't be able to run beaker and rspec puppet code since beaker inherently requires rspec 3 and rspec-puppet forces rspec 2.  I see this as a big problem.
 
https://github.com/electrical/rspec-puppet/commit/83146c2010ee454a2ac1ef38a3fb797a2337915d

**Comment by dfarrell07** [comment]
Having `rspec-puppet` pinned to `rspec ~> 2.0` causes a somewhat deep issue that I've been beating my head against all day.

tl;dr is that `serverspec` versions in `~> 2.0` have a fix related to systemd that's required for checking if services are enabled (ie using Beaker) in Fedora 21. However, `serverspec ~> 2.0` needs `rspec 3.0.0` and `rspec-puppet` needs `rspec ~> 2.0`, so I'm stuck using `serverspec 1.16.0`, so checking if services are enabled using Beaker against F21 is broken.

```
Bundler could not find compatible versions for gem "rspec":
  In Gemfile:
    rspec-puppet (>= 0) ruby depends on
      rspec (~> 2.0) ruby

    serverspec (~> 2.0) ruby depends on
      rspec (3.0.0)
```

**Comment by hirocaster** [comment]
:+1: 

rspec 2.x version is "old version".

rspec version up 3.x from 2.x is easy, use the [transpec](https://github.com/yujinakayama/transpec)
other gems version uped ...

```
Resolving dependencies...
Bundler could not find compatible versions for gem "rspec":  In Gemfile:
    rspec-puppet (>= 0) ruby depends on
      rspec (~> 2.0) ruby
    beaker-rspec (>= 0) ruby depends on
      rspec (3.1.0)
```

**Comment by dfarrell07** [comment]
I've been extensively using/abusing the branch this PR is based on all week with no issues, btw. imho, it's a safe change that should be merged ASAP.

**Comment by raphink** [comment]
It's currently not possible to have both `rspec-puppet` and `serverspec` installed together because of this issue.

Could you switch to rspec 3 please?

**Comment by carlossg** [comment]
:+1:

**Comment by dfarrell07** [comment]
@rodjek - Is there any reason this hasn't been merged that we can help address?

**Comment by cmurphy** [comment]
The puppetlabs modules are experiencing the same issue that @dfarrell07 described, so we're stuck on rspec-puppet 1.0.1 until that dependency changes.

@jeffmccune explained in https://github.com/rodjek/rspec-puppet/commit/a5540918c983011adc2620a210717484e11cce43 that there /may/ be cases where rspec 3 isn't compatible with rspec-puppet 2, and has explained elsewhere that it is typical for rspec-* gems to fix the rspec dependency to avoid compatibility conflicts. I would note that in practice, at least in our module testing, rspec-puppet 1.x has worked fine with rspec 3.x, though that may not be the case for everyone.

**Comment by tayzlor** [comment]
:+1:

**Comment by logicminds** [comment]
@cmurphy I think having this dependency on 2 is causing more issues than having it unpinned to a specific version would cause.  Many of us have used rspec 3 for about a year with no issues and then the dependency of 2 was forced upon us which no longer allows us to use beaker and rspec-puppet together.  Besides this is not about forcing rspec 3 on anybody.  What this is about is about removing the dependency on a specific version of rspec-puppet.  If the user wishes to use rspec-2 they can, and if the user wants rspec-3 they can also use rspec-3  (mandated by beaker/serverspec).  This pin to rspec-2 was in https://github.com/rodjek/rspec-puppet/commit/a5540918c983011adc2620a210717484e11cce43 at which time it worked great until this was introduced.

Furthermore, we can explicitly still allow the should syntax.

```ruby
RSpec.configure do |config|
  config.expect_with :rspec do |c|
    c.syntax = [:should, :expect]
  end
end

```
https://relishapp.com/rspec/rspec-expectations/v/3-2/docs/syntax-configuration

Additionally there is now a tool to convert rspec2 syntax to rspec3 syntax
http://yujinakayama.me/transpec/

**Comment by dfarrell07** [comment]
Note that the `rspec-puppet` project in general isn't merging PRs quickly at the moment ([last one](https://github.com/rodjek/rspec-puppet/commit/8459e14807977244c00bdbcf190062c529b63474) was Dec 2nd, almost 3 months ago). I think the issue is project-wide, not specific to this PR.

I pinged the owner of the repo (@rodjek) [in this issue](https://github.com/rodjek/rspec-puppet/pull/249#issuecomment-74083896) as well as [on Twitter](https://twitter.com/dfarrell07/status/568797298320412672), but didn't get a response. Since @jeffmccune was the last person to merge a PR, I also [pinged him on Twitter](https://twitter.com/dfarrell07/status/568834925954265088). He [responded](https://twitter.com/0xEFF/status/569078509965959168) and said he would take a look at the PR backlog ASAP.

**Comment by cmurphy** [comment]
@logicminds :+1:

**Comment by electrical** [comment]
@dfarrell07 thanks for chasing it :-)

**Comment by dfarrell07** [comment]
I raised the [lack of maintenance concern](https://github.com/rodjek/rspec-puppet/pull/249#issuecomment-75811101) on the [#puppet-dev IRC channel yesterday](http://irc-logs.puppetlabs.com/puppetdev/%23puppet-dev-2015-03-03.log.html#t2015-03-03T11:23:25).

[nibalizer suggested moving rspec-puppet](http://irc-logs.puppetlabs.com/puppetdev/%23puppet-dev-2015-03-03.log.html#t2015-03-03T11:35:35) to the [Puppet Community GitHub org](https://github.com/puppet-community), which would apparently provide a larger set of folks for reviewing PRs.

[It was suggested](http://irc-logs.puppetlabs.com/puppetdev/%23puppet-dev-2015-03-03.log.html#t2015-03-03T11:45:18) that I hang around and talk to rodjek about it on IRC, but from what I can see of the [day's logs](http://irc-logs.puppetlabs.com/puppetdev/%23puppet-dev-2015-03-03.log.html) he never joined that IRC channel.

As [I mentioned](http://irc-logs.puppetlabs.com/puppetdev/%23puppet-dev-2015-03-03.log.html#t2015-03-03T11:42:22) to nibalizer, my plan is to either talk to rodjek soon on IRC or raise a dedicated lack-of-maintenance issue about  against rspec-puppet and drop links to it until someone pays attention. Any help would be appreciated. :)

**Comment by raphink** [comment]
I talked about this with @daenney recently, and he was going to write @rodjek about it. I already transferred camptocamp's puppet-lint modules to the puppet community group, and we thought it might make sense to request the adoption of puppet-lint and rspec-puppet, since they are both important community projects that would benefit from a shared maintenance.

**Comment by dfarrell07** [comment]
> I talked about this with @daenney recently, and he was going to write @rodjek about it.

Awesome, great to know that people are thinking about it. Thanks! :)

> since they are both important community projects that would benefit from a shared maintenance.

+1, makes sense to me.

**Comment by logicminds** [comment]
This all sounds good.  What exactly is the puppet community project?

**Comment by daenney** [comment]
This: https://github.com/puppet-community

**Comment by raphink** [comment]
@logicminds The puppet community group comprises main contributors to the puppet ecosystem, including people from Puppet Labs, Spotify, Camptocamp, etc. The idea is to co-maintain projects related to puppet.

**Comment by raphink** [comment]
See http://puppet-community.github.io/

**Comment by dfarrell07** [comment]
I opened an issue (#260) against rspec-puppet that's dedicated to the Puppet Community transfer topic. I suggest we attempt to let this PR thread re-focus on the rspec3 update.

Thanks for helping to improve rspec-puppet, everyone. :)

**Comment by dfarrell07** [comment]
Thank you @jeffmccune! :+1:

**Comment by jeffmccune** [comment]
The last time I was working on rpsec-puppet, I got stuck trying to push to rubygems.org.  Rodjek has just recently given me the ability to push to rubygems, so I'm going to try and get that release out the door now.

There's definitely some great long term maintenance issues that we still need to sort out, I'm not going to be able to handle this myself, but I can at least help release some of the pressure that's built up which is what I'll be doing tonight and tomorrow.

**Comment by logicminds** [comment]
Thanks for putting the time in @jeffmccune.  I know rspec-puppet internals fairly well so if there's any way I can help out let me know.

**Comment by petems** [comment]
Woop woop! :+1:

**Comment by dfarrell07** [comment]
I've [successfully move back](https://github.com/dfarrell07/puppet-opendaylight/issues/16) to official rodjek/rspec-puppet from @logicminds' pre-merge fork. Everything's working well, as expected. Thanks again all!

```
[~/puppet-opendaylight]$ git diff | cat
diff --git a/Gemfile b/Gemfile
index a6f4f2e..2658802 100644
--- a/Gemfile
+++ b/Gemfile
@@ -3,7 +3,7 @@ source "https://rubygems.org"
 group :test do
   gem "rake"
   gem "puppet", ENV['PUPPET_VERSION'] || '~> 3.7.0'
-  gem "rspec-puppet", :git => 'https://github.com/logicminds/rspec-puppet'
+  gem "rspec-puppet"
   gem "puppetlabs_spec_helper"
   gem "metadata-json-lint"
 end
[~/puppet-opendaylight]$ bundle update
#snip
Installing rspec-puppet 2.0.1 (was 2.0.0)
#snip
[~/puppet-opendaylight]$ bundle exec rake test
#snip
Finished in 7.61 seconds (files took 0.53749 seconds to load)
398 examples, 0 failures
#snip
```
